gateway/src/shard/processor/impl.rs

```diff
@@ -34,8 +35,184 @@ use twilight_model::gateway::{
         identify::{Identify, IdentifyInfo, IdentifyProperties},
         resume::Resume,
     },
+    GatewayIntents,
 };
-use url::Url;
+use url::{ParseError as UrlParseError, Url};
+
+#[cfg(not(feature = "simd-json"))]
+use serde_json::Error as JsonError;
+#[cfg(feature = "simd-json")]
+use simd_json::Error as JsonError;
+
+/// Connecting to the gateway failed.
+#[derive(Debug)]
+pub enum ConnectingError {
+    Establishing { source: TungsteniteError },
+    ParsingUrl { source: UrlParseError, url: String },
+}
+
+impl Display for ConnectingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Self::Establishing { source } => Display::fmt(source, f),
+            Self::ParsingUrl { source, url } => f.write_fmt(format_args!(
+                "the gateway url `{}` is invalid: {}",
+                url, source,
+            )),
+        }
+    }
+}
+
+impl Error for ConnectingError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Self::Establishing { source } => Some(source),
+            Self::ParsingUrl { source, .. } => Some(source),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum GatewayEventParsingError {
+    /// Deserializing the GatewayEvent payload from JSON failed.
+    Deserializing {
+        /// Reason for the error.
+        source: JsonError,
+    },
+    /// The payload received from Discord was an invalid structure.
+    ///
+    /// The payload was either invalid JSON or did not contain the necessary
+    /// "op" key in the object.
+    PayloadInvalid,
+}
+
+impl Display for GatewayEventParsingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Self::Deserializing { source } => Display::fmt(source, f),
+            Self::PayloadInvalid => f.write_str("payload is an invalid json structure"),
+        }
+    }
+}
+
+impl Error for GatewayEventParsingError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Self::Deserializing { source } => Some(source),
+            Self::PayloadInvalid => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum ProcessError {
+    /// A close message tried to be sent but the receiving half was dropped.
+    /// This typically means that the shard is shutdown.
+    SendingClose {
+        /// Reason for the error.
+        source: TrySendError<Message>,
+    },
+    SessionSend {
```

Comment:
Should there be a comment on this variant too?

Reply:
This enum isn't exposed to the public API but I've documented it for development documentation purposes